src/Nethermind/Nethermind.Network/PeerManager.MorePeersNeeded.cs

```diff
@@ -0,0 +1,351 @@
+﻿//  Copyright (c) 2021 Demerzel Solutions Limited
+//  This file is part of the Nethermind library.
+// 
+//  The Nethermind library is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+// 
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Nethermind.Core.Crypto;
+using Nethermind.Stats.Model;
+// ReSharper disable InconsistentNaming
+
+namespace Nethermind.Network;
+
+public partial class PeerManager
+{
+    private class MorePeersNeeded : PeeringEvent
```

Comment:
Wow this part needs redoing/refactoring so much.
Also I think this can be quite long, won't it starve other events?